bentoml/_internal/monitoring/api.py

```diff
@@ -0,0 +1,232 @@
+
+if TYPE_CHECKING:
+    from ..types import JSONSerializable
+
+    DT = t.TypeVar("DT")
+    MT = t.TypeVar("MT", bound="Monitor[t.Any]")
+
+logger = logging.getLogger(__name__)
+
+BENTOML_MONITOR_ROLES = {"feature", "prediction", "target"}
+BENTOML_MONITOR_TYPES = {"numerical", "categorical", "numerical_sequence"}
+
+MONITORS: dict[str, Monitor[t.Any]] = {}  # cache of monitors
+
+
+class Monitor(t.Generic[DT]):
+    def __init__(self, name: str):
+        pass
+
+    def start_record(self) -> None:
+        pass
+
+    def stop_record(self) -> None:
+        pass
+
+    def export_schema(self) -> JSONSerializable:
+        pass
+
+    def export_data(self) -> JSONSerializable:
+        pass
+
+    def log(self, data: DT, name: str, role: str, data_type: str) -> None:
+        pass
+
+    def log_batch(
+        self, data_batch: t.Iterable[DT], name: str, role: str, data_type: str
+    ) -> None:
+        pass
+
+    def log_table(
+        self,
+        data: t.Iterable[t.Iterable[DT]],
+        schema: dict[str, str],
+    ) -> None:
+        pass
+
+
+class BentoMLDefaultMonitor(Monitor["JSONSerializable"]):
+    PRESERVED_COLUMNS = (COLUMN_TIME, COLUMN_RID) = ("timestamp", "request_id")
+
+    @inject
+    def __init__(
+        self,
+        name: str,
+        log_config: dict[str, t.Any] = Provide[BentoMLContainer.monitoring_log_config],
+    ) -> None:
+        self.name = name
+        logging.config.dictConfig(log_config)
+        self.data_logger = logging.getLogger("bentoml_monitor_data")
+        self.schema_logger = logging.getLogger("bentoml_monitor_schema")
+        self._is_first_record = True
+        self._is_first_column = False
+        self._schema: list[dict[str, str]] = []
+        self._columns: dict[
+            str,
+            collections.deque[JSONSerializable],
+        ] = collections.defaultdict(collections.deque)
+
+    def start_record(self) -> None:
+        """
+        Start recording data. This method should be called before logging any data.
+        """
+        self._is_first_column = True
+
+    def stop_record(self) -> None:
+        """
+        Stop recording data. This method should be called after logging all data.
+        """
+        if not self._is_first_record:
+            self.export_schema()
+            self._is_first_record = False
+
+        if self._is_first_column:
+            logger.warning("No data logged in this record. Will skip this record.")
+        else:
+            self.export_data()
+
+    def export_schema(self):
+        """
+        Export schema of the data. This method should be called after all data is logged.
+        """
+        self.schema_logger.info(self._schema)
+
+    def export_data(self):
+        """
+        Export data. This method should be called after all data is logged.
+        """
+        assert (
+            len(set(len(q) for q in self._columns.values())) == 1
+        ), "All columns must have the same length"
```

Comment:
Wouldn't this mean that the time and request id columns must also have the same length?

Reply:
yeah. It's expected